src/main/java/cloud/eppo/BanditEvaluator.java

```diff
@@ -0,0 +1,167 @@
+                  actionScore +=
+                      scoreContextForCoefficients(
+                          actionAttributes.getNumericAttributes(),
+                          banditCoefficients.getActionNumericCoefficients());
+                  actionScore +=
+                      scoreContextForCoefficients(
+                          actionAttributes.getCategoricalAttributes(),
+                          banditCoefficients.getActionCategoricalCoefficients());
+                  actionScore +=
+                      scoreContextForCoefficients(
+                          subjectAttributes.getNumericAttributes(),
+                          banditCoefficients.getSubjectNumericCoefficients());
+                  actionScore +=
+                      scoreContextForCoefficients(
+                          subjectAttributes.getCategoricalAttributes(),
+                          banditCoefficients.getSubjectCategoricalCoefficients());
+
+                  return actionScore;
+                }));
+  }
+
+  private static double scoreContextForCoefficients(
+      Attributes attributes, Map<String, ? extends BanditAttributeCoefficients> coefficients) {
+
+    double totalScore = 0.0;
+
+    for (BanditAttributeCoefficients attributeCoefficients : coefficients.values()) {
+      EppoValue contextValue = attributes.get(attributeCoefficients.getAttributeKey());
+      // The coefficient implementation knows how to score
+      double attributeScore = attributeCoefficients.scoreForAttributeValue(contextValue);
+      totalScore += attributeScore;
+    }
+
+    return totalScore;
+  }
+
+  private static Map<String, Double> weighActions(
+      Map<String, Double> actionScores, double gamma, double actionProbabilityFloor) {
+    Double highestScore = null;
+    String highestScoredAction = null;
+    for (Map.Entry<String, Double> actionScore : actionScores.entrySet()) {
+      if (highestScore == null
+          || actionScore.getValue() > highestScore
+          || actionScore
+                  .getValue()
+                  .equals(highestScore) // note: we break ties for scores by action name
+              && actionScore.getKey().compareTo(highestScoredAction) < 0) {
+        highestScore = actionScore.getValue();
+        highestScoredAction = actionScore.getKey();
+      }
+    }
+
+    // Weigh all the actions using their score
+    Map<String, Double> actionWeights = new HashMap<>();
+    double totalNonHighestWeight = 0.0;
+    for (Map.Entry<String, Double> actionScore : actionScores.entrySet()) {
+      if (actionScore.getKey().equals(highestScoredAction)) {
+        // The highest scored action is weighed at the end
+        continue;
+      }
+
+      // Compute weight (probability)
+      double unboundedProbability =
+          1 / (actionScores.size() + (gamma * (highestScore - actionScore.getValue())));
+      double boundedProbability = Math.max(unboundedProbability, actionProbabilityFloor);
```

Comment:
In the other SDKs, we divide [actionProbabilityFloor by the number of actions to weigh;](https://github.com/Eppo-exp/python-sdk/blob/70595d1c00851f383095c777de6480bbb882b480/eppo_client/bandit.py#L167)

[Comment 🧵 ](https://github.com/Eppo-exp/python-sdk/pull/38/files#r1619789521)

Reply:
Yikes! GREAT catch 🎉  Dividing by actions is what we should do. 

This logic was part of what was copied over from the original Java implementation, before we changed up our strategy, and I failed to update it.

What this does mean is that we don't have a test enforcing that logic. It may be hard to put one together, but I'll see if I can. 🤔

Reply:
The reason why this doesn't get teased out in a test is because we're using small min probability floors and typically a gamma of 1 with only a couple of actions. With some tweaking of scoring parameters, should be able to create a test case that pushes it over the edge

Reply:
PR for test cases that cover this (and more!) here: https://github.com/Eppo-exp/sdk-test-data/pull/52